python/lsst/cp/pipe/ptc.py

```diff
@@ -367,11 +434,269 @@ def runDataRef(self, dataRef, visitPairs):
+            gain = fit.getGain()
+            dataset.gain.setdefault(amp, []).append(gain)
+            dataset.noise.setdefault(amp, []).append(np.sqrt(fit.getRon()))
+            dataset.signal.setdefault(amp, []).append(meanVecFinal)
+            dataset.var.setdefault(amp, []).append(varVecFinal)
+            dataset.varModel.setdefault(amp, []).append(varVecModel)
+            dataset.signalAdu.setdefault(amp, []).append(meanVecFinal/gain)
+            dataset.varAdu.setdefault(amp, []).append(varVecFinal/(gain**2))
+            dataset.varModelAdu.setdefault(amp, []).append(varVecModel/gain**2)
+            dataset.a00.setdefault(amp, []).append(fit.getA()[0, 0])
+            dataset.b00.setdefault(amp, []).append(fit.getB()[0, 0])
+
+        return dataset
+
+    def computeCovariancesAstier(self, dataRef, visitPairs, detector):
+        """Iterate over pairs of flats to calculate full covariances from the difference image.
+        This follows the treatment in Astier+19
+
+        Parameters
+        ----------
+        dataRef : list of lsst.daf.persistence.ButlerDataRef
+            dataRef for the detector for the visits to be fit.
+
+        visitPairs : `iterable` of `tuple` of `int`
+            Pairs of visit numbers to be processed together
+
+        detector : `lsst.afw.cameraGeom.Detector`
+            Detector object
+
+        Returns
+        -------
+        covariancesWithTags : `numpy.recarray`
+            Covariance measurements.
+        """
+        tupleRecords = []
+        allTags = []
+
+        self.log.info('Measuring Astier covariances using %s visits for detector %s' % (visitPairs,
+                                                                                        detector.getId()))
+
+        for (v1, v2) in visitPairs:
+            # Perform ISR on each exposure
+            dataRef.dataId['expId'] = v1
+            exp1 = self.isr.runDataRef(dataRef).exposure
+            dataRef.dataId['expId'] = v2
+            exp2 = self.isr.runDataRef(dataRef).exposure
+            del dataRef.dataId['expId']
+
+            checkExpLengthEqual(exp1, exp2, v1, v2, raiseWithMessage=True)
+            expTime = exp1.getInfo().getVisitInfo().getExposureTime()
+
+            tupleRows = []
+            for ampNumber, amp in enumerate(detector):
+                # covs: (i, j, var, cov, npix)
+                mu1, mu2, covs = self.getCovariancesAstier(exp1, exp2, region=amp.getBBox())
+                tupleRows += [(mu1, mu2) + cov + (ampNumber, expTime, amp.getName()) for cov in covs]
+                tags = ['mu1', 'mu2', 'i', 'j', 'var', 'cov', 'npix', 'ext', 'expTime', 'ampName']
+            allTags += tags
+            tupleRecords += tupleRows
+        covariancesWithTags = np.core.records.fromrecords(tupleRecords, names=allTags)
+        return covariancesWithTags
+
+    def getCovariancesAstier(self, exposure1, exposure2, region=None):
```

Comment:
This method also shares a lot of code with measureMeanVarPair.  Can they be unified into one method?
(It seems like that would invalidate my previous comment, moving the config option check into that unified method, and out of the run() method.)